pkg/nfd-master/nfd-master.go

```diff
@@ -432,10 +446,116 @@ func (m *nfdMaster) SetLabels(c context.Context, r *pb.SetLabelsRequest) (*pb.Se
 			klog.Errorf("failed to advertise labels: %v", err)
 			return &pb.SetLabelsReply{}, err
 		}
+
+		// set taints
+		switch m.args.EnableTaints {
+		case true:
+			labelReply, err := m.setTaints(c, crt, r.NodeName)
+			if err != nil {
+				return labelReply, err
+			}
+			// Call setTaints() with empty list of taints even though the feature
+			// flag is disabled. This ensures that we delete NFD owned stale taints
+			// when flag got turned off.
+		case false:
+			labelReply, err := m.setTaints(c, emptyTaint, r.NodeName)
+			if err != nil {
+				return labelReply, err
+			}
+		}
 	}
 	return &pb.SetLabelsReply{}, nil
 }
 
+// setTaints sets note taints and annotations based on the taints passed via
+// nodeFeatureRule custom resorce. If empty list of taints is passed, currently
+// NFD owned taints and annotations are removed from the node.
+func (m *nfdMaster) setTaints(c context.Context, taints []corev1.Taint, nodeName string) (*pb.SetLabelsReply, error) {
+	var nodeTaints, annotationTaints []corev1.Taint
+	var val string
+	taintStrs := []string{}
+
+	err := authorizeClient(c, m.args.VerifyNodeName, nodeName)
+	if err != nil {
+		return &pb.SetLabelsReply{}, err
+	}
+
+	cli, err := m.apihelper.GetClient()
+	if err != nil {
+		return &pb.SetLabelsReply{}, err
+	}
+
+	// fetch the node object
+	node, err := m.apihelper.GetNode(cli, nodeName)
+	if err != nil {
+		return &pb.SetLabelsReply{}, err
+	}
+
+	// update the taint annotation
+	annotation := node.GetAnnotations()
```

Comment:
super nit: many annotations in this var so rename to `annotations`(?)


```suggestion
	annotations := node.GetAnnotations()
```